File: flutter/lib/core/services/auth_service.dart
import 'dart:convert';
import '../models/auth_models.dart';
import '../models/api_response.dart';
import 'api_service.dart';

class AuthService {
  final ApiService _api = ApiService();

  Future<ApiResponse<LoginResult>> login(LoginRequest request) async {
    return _api.post<LoginResult>(
      '/auth/login',
      data: request.toJson(),
      fromData: (data) => LoginResult.fromJson(data),
    );
  }

  AuthUser? decodeToken(String token) {
    try {
      final parts = token.split('.');
      if (parts.length != 3) return null;
      var payload = parts[1];
      payload = base64.normalize(payload);
      final decoded = utf8.decode(base64.decode(payload));
      final json = jsonDecode(decoded) as Map<String, dynamic>;

      final roles = json['role'] is List
          ? List<String>.from(json['role'])
          : json['role'] is String
              ? [json['role'] as String]
              : <String>[];

      return AuthUser(
        id: json['sub'] ?? '',
        tenantId: json['tenantId'] ?? '',
        name: json['unique_name'] ?? '',
        surname: '',
        email: json['email'] ?? '',
        roles: roles,
      );
    } catch (_) {
      return null;
    }
  }

  bool isTokenExpired(String token) {
    try {
      final parts = token.split('.');
      if (parts.length != 3) return true;
      var payload = parts[1];
      payload = base64.normalize(payload);
      final decoded = utf8.decode(base64.decode(payload));
      final json = jsonDecode(decoded) as Map<String, dynamic>;
      final exp = json['exp'] as int;
      return DateTime.fromMillisecondsSinceEpoch(exp * 1000).isBefore(DateTime.now());
    } catch (_) {
      return true;
    }
  }
}

File: desktop/BeautyEstiva.Desktop/BeautyEstiva.Desktop/Services/AuthService.cs
using BeautyEstiva.Desktop.Helpers;
using BeautyEstiva.Desktop.Models;

namespace BeautyEstiva.Desktop.Services;

public class AuthService : IAuthService
{
    private readonly IApiService _apiService;
    private readonly ITokenStorageService _tokenStorage;

    public AuthUser? CurrentUser { get; private set; }
    public bool IsAuthenticated => CurrentUser != null;
    public event Action? AuthStateChanged;

    public AuthService(IApiService apiService, ITokenStorageService tokenStorage)
    {
        _apiService = apiService;
        _tokenStorage = tokenStorage;
    }

    public async Task<ApiResponse<LoginResult>> LoginAsync(LoginRequest request)
    {
        var response = await _apiService.PostAsync<LoginResult>("/auth/login", request);

        if (response.Success && response.Data != null)
        {
            var result = response.Data;
            _apiService.SetToken(result.Token);

            var payload = JwtHelper.DecodeToken(result.Token);
            CurrentUser = new AuthUser
            {
                Id = payload?.Sub ?? "",
                TenantId = payload?.TenantId ?? "",
                Name = result.Name,
                Surname = result.Surname,
                Email = result.Email,
                Roles = result.Roles
            };

            AuthStateChanged?.Invoke();
        }

        return response;
    }

    public void Logout()
    {
        _apiService.SetToken(null);
        CurrentUser = null;
        AuthStateChanged?.Invoke();
    }

    public bool TryRestoreSession()
    {
        var token = _tokenStorage.GetToken();
        if (string.IsNullOrEmpty(token))
            return false;

        if (JwtHelper.IsTokenExpired(token))
        {
            _tokenStorage.ClearToken();
            return false;
        }

        var payload = JwtHelper.DecodeToken(token);
        if (payload == null)
        {
            _tokenStorage.ClearToken();
            return false;
        }

        _apiService.SetToken(token);

        var roles = JwtHelper.ExtractRoles(payload);
        CurrentUser = new AuthUser
        {
            Id = payload.Sub,
            TenantId = payload.TenantId,
            Name = payload.UniqueName,
            Email = payload.Email,
            Roles = roles
        };

        AuthStateChanged?.Invoke();
        return true;
    }
}
